File: Medium/33_Search_in_Rotated_Sorted_Array/Solution.py
from typing import Optional, List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while (left <= right):
            mid = (left + right) // 2

            if (nums[mid] == target):
                return mid
            
            if (nums[mid] >= nums[left]):
                if (nums[left] <= target < nums[mid]):
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                if (nums[mid] < target <= nums[right]):
                    left = mid + 1
                else:
                    right = mid - 1
        
        return -1



if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    nums1 = [4,5,6,7,0,1,2]
    target1 = 0
    print(solution.search(nums1, target1))

    # test cases 2
    nums1 = [4,5,6,7,0,1,2]
    target1 = 3
    print(solution.search(nums1, target1))

    # test cases 3
    nums1 = [1]
    target1 = 0
    print(solution.search(nums1, target1))

File: Medium/33_Search_in_Rotated_Sorted_Array/Solution.java



class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (nums[mid] == target) {
                return mid;
            }

            // Check if the left half is sorted
            if (nums[left] <= nums[mid]) {
                if (target >= nums[left] && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } 
            // Otherwise, the right half must be sorted
            else {
                if (target > nums[mid] && target <= nums[right]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        int[] nums1 = {4, 5, 6, 7, 0, 1, 2};
        int target1 = 0;
        System.out.println(solution.search(nums1, target1)); // Expected output: 4

        // test cases 2
        int[] nums2 = {4, 5, 6, 7, 0, 1, 2};
        int target2 = 3;
        System.out.println(solution.search(nums2, target2)); // Expected output: -1

        // test cases 3
        int[] nums3 = {1};
        int target3 = 0;
        System.out.println(solution.search(nums3, target3)); // Expected output: -1
        
    }
}